recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py

```diff
@@ -0,0 +1,157 @@
+import argparse
+from bson.binary import Binary
+import datetime
+import difflib
+import mmap
+import os
+import pymongo
+import re
+import subprocess
+import sys
+
+class DB:
+    mongo_db_name = 'test'
+    mongo_db_collection_name = 'rtos_dmesg_logs'
+
+    def __init__(self, server, user, password):
+        mongo_client = pymongo.MongoClient("mongodb://{user}:{pwd}@{server}".format(
+            user=args.user, pwd=args.password, server=args.server
+            ),
+            w=1)
+        db = mongo_client[self.mongo_db_name]
+        self.dmesg_logs_collection = db[self.mongo_db_collection_name]
+
+    def insert(self, data):
+        return self.dmesg_logs_collection.insert_one(data)
+
+    def find(self, query):
+        return self.dmesg_logs_collection.find(query)
+
+def run_cmd(cmd):
+    output = subprocess.check_output(cmd).strip()
+    return output.decode('utf-8')
+
+def get_device_desc():
+    cmdoutput = run_cmd(['fw_printenv', 'DeviceDesc'])
+    return cmdoutput.split('=')[1]
+
+def get_kernel_version():
+    return run_cmd(['uname', '-r'])
+
+def get_architecture():
+    return run_cmd(['uname', '-m'])
+
+def get_os_version():
+    with open('/etc/os-release', 'rb') as file, mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) as mfile:
+        build_id = re.search(br'BUILD_ID=\"(.*)\"', mfile)
+        return build_id.group(1).decode()
+
+def is_next_kernel():
+    kernel_version = get_kernel_version()
+    return kernel_version.endswith('-next')
+
+def get_dmesg_log():
+    dmesg_log = run_cmd(['dmesg'])
+    return dmesg_log
+
+def upload_log(db, dmesg_log):
+    data = {}
+    data['kernel_version'] = get_kernel_version()
+    data['kernel_type'] = 'next' if is_next_kernel() else 'current'
+    data['device_desc'] = get_device_desc()
+    data['architecture'] = get_architecture()
+    data['os_version'] = get_os_version()
+    data['date'] = str(datetime.datetime.now())
+
+    header = ''
+    for key, val in data.items():
+        header = header + '# {}: {}\n'.format(key, val)
+
+    data['dmesg_log'] = header + dmesg_log
+
+    record = db.insert(data)
+    print('INFO: Uploaded current dmesg log from {} with _id {}'.format(data['date'], record.inserted_id))
+
+def strip_headers(dmesg_log):
+    output = ''
+    for line in dmesg_log.splitlines():
+        if not line.startswith('#'):
+            output = output + line + '\n'
+    return output
+
+def get_old_dmesg_log(db):
+    query = {}
+    query['kernel_version'] = {'$ne': get_kernel_version()}
+    query['kernel_type'] = 'next' if is_next_kernel() else 'current'
+    query['device_desc'] = get_device_desc()
+    results = db.find(query).sort('date', pymongo.DESCENDING).limit(1)
+    if not results:
+        return None
+
+    for result in results:
```

Comment:
This for loop reads weird to me since we're returning with the first 'result' in 'results' i.e. do we need the loop?

Reply:
`results` is a generator and if we don't want to use the loop method can use `next(results)` but this fails if there are no results - so it would require putting a try catch block which ends up complicating the logic even more. This seemed the simplest option to me.

But if someone knows a better way, I'm happy to change this.

Reply:
OK, I think it's fine.

Reply:
Modified the implementation to account for the new 'prefer same major.minor' logic and got rid of this loop.